<h1 style="font-family:courier new;">Data Science and Machine Learning IBM Capstone Project</h1>

![NYC 311](https://lh3.googleusercontent.com/frrVTacZfKCUIapOpvbtIW6BaqwwKgEyiZ_84ad4hWg1jF8I5cUirv0GLL8rFZtdCRJQ)

## Problem Statement

The people of New York use the __*311 system*__ to report complaints about the non-emergency problems to local authorities. Various agencies in New York are assigned these problems. The __*Department of Housing Preservation and Development of New York City*__ is the agency that processes 311 complaints that are related to housing and buildings.

In the last few years, the number of 311 complaints to the Department of Housing Preservation and Development has increased significantly. Although these complaints are not necessarily urgent, the large volume of complaints and the sudden increase is impacting the overall efficiency of operations of the agency.

The Department of Housing Preservation and Development has raised the following questions to which they need answers supported by both data and analytics.
1. Which type of complaint should the Department of Housing Preservation and Development of New York City focus on first?
2. Should the Department of Housing Preservation and Development of New York City focus on any particular set of boroughs, ZIP codes, or street (where the complaints are severe) for the specific type of complaints you identified in response to Question 1?

## Datasets

Two datasets from the Department of Housing Preservation and Development of New York City are used to address their problems.
- ### 311 complaint dataset

     This dataset is available at **https://data.cityofnewyork.us/Social-Services/311-Service-Requests-from-2010-to-Present/erm2-nwe9**. Download only the data that is related to the Department of Housing Preservation and Development.
The data can be download using this link: **https://cocl.us/311_NYC_Dataset**. The downloaded data will have complaints submitted between 2010 and 2020.

- ### PLUTO dataset for housing

     This dataset for housing can be accessed from **https://data.cityofnewyork.us/City-Government/Primary-Land-Use-Tax-Lot-Output-PLUTO-/xuk2-nczf**. 
Use the PLUTO dataset URL to download the zip file: **https://www1.nyc.gov/assets/planning/download/zip/data-maps/open-data/nyc_pluto_18v1.zip**.
     Extract the file.It has five CSV files for the five New York city boroughs: Bronx, Brooklyn, Manhattan, Queens, and Staten Island.

## Data Wrangling

The data downloaded is imported into a python dataframe using pandas library. 

In [1]:
import pandas as pd

In [2]:
nyc311=pd.read_csv(r"C:\Users\Josna Baby\Downloads\311_Service_Requests_from_2010_to_Present_min.csv")

In [3]:
nyc311.head()

Unnamed: 0  Unique Key            Created Date Closed Date  \
0           0    45531130  02/02/2020 06:09:17 AM         NaN   
1           1    45529784  02/02/2020 02:15:24 PM         NaN   
2           2    45527528  02/02/2020 02:27:41 AM         NaN   
3           3    45530329  02/02/2020 12:13:18 PM         NaN   
4           4    45528814  02/02/2020 01:59:44 PM         NaN   

         Complaint Type         Location Type  Incident Zip  \
0        HEAT/HOT WATER  RESIDENTIAL BUILDING       10019.0   
1  UNSANITARY CONDITION  RESIDENTIAL BUILDING       11204.0   
2        HEAT/HOT WATER  RESIDENTIAL BUILDING       11372.0   
3        HEAT/HOT WATER  RESIDENTIAL BUILDING       10458.0   
4             APPLIANCE  RESIDENTIAL BUILDING       11209.0   

          Incident Address         Street Name Address Type             City  \
0     426 WEST   52 STREET    WEST   52 STREET      ADDRESS         NEW YORK   
1           1751 67 STREET           67 STREET      ADDRESS         BROOKLYN   
2          87-15 37 AVENUE           37 AVENUE      ADDRESS  Jackson Heights   
3  2405 SOUTHERN BOULEVARD  SOUTHERN BOULEVARD      ADDRESS            BRONX   
4            223 78 STREET           78 STREET      ADDRESS         BROOKLYN   

  Status                             Resolution Description    Borough  \
0   Open  The following complaint conditions are still o...  MANHATTAN   
1   Open  The following complaint conditions are still o...   BROOKLYN   
2   Open  The following complaint conditions are still o...     QUEENS   
3   Open  The following complaint conditions are still o...      BRONX   
4   Open  The following complaint conditions are still o...   BROOKLYN   

    Latitude  Longitude  
0  40.765132 -73.988993  
1  40.618484 -73.992673  
2  40.750269 -73.879432  
3  40.853773 -73.881558  
4  40.629745 -74.030533

In [4]:
nyc311.tail()

Unnamed: 0  Unique Key            Created Date  \
6019838     6019838    44063685  10/15/2019 10:35:45 AM   
6019839     6019839    44063692  10/15/2019 06:09:28 AM   
6019840     6019840    44063724  10/15/2019 10:54:45 AM   
6019841     6019841    44063726  10/15/2019 02:07:34 PM   
6019842     6019842    44063737  10/15/2019 11:29:01 AM   

                    Closed Date  Complaint Type         Location Type  \
6019838  10/16/2019 05:55:02 PM  HEAT/HOT WATER  RESIDENTIAL BUILDING   
6019839  10/15/2019 04:12:58 PM  HEAT/HOT WATER  RESIDENTIAL BUILDING   
6019840  10/16/2019 06:14:29 PM  HEAT/HOT WATER  RESIDENTIAL BUILDING   
6019841  10/16/2019 07:10:19 PM  HEAT/HOT WATER  RESIDENTIAL BUILDING   
6019842  10/15/2019 09:19:45 PM   PAINT/PLASTER  RESIDENTIAL BUILDING   

         Incident Zip         Incident Address         Street Name  \
6019838       10029.0     327 EAST  108 STREET    EAST  108 STREET   
6019839       10461.0  3555 BRUCKNER BOULEVARD  BRUCKNER BOULEVARD   
6019840       10034.0       165 SHERMAN AVENUE      SHERMAN AVENUE   
6019841       10467.0    66 WEST GUN HILL ROAD  WEST GUN HILL ROAD   
6019842       10009.0             143 AVENUE D            AVENUE D   

        Address Type      City  Status  \
6019838      ADDRESS  NEW YORK  Closed   
6019839      ADDRESS     BRONX  Closed   
6019840      ADDRESS  NEW YORK  Closed   
6019841      ADDRESS     BRONX  Closed   
6019842      ADDRESS  NEW YORK  Closed   

                                    Resolution Description    Borough  \
6019838  The complaint you filed is a duplicate of a co...  MANHATTAN   
6019839  The Department of Housing Preservation and Dev...      BRONX   
6019840  The complaint you filed is a duplicate of a co...  MANHATTAN   
6019841  The Department of Housing Preservation and Dev...      BRONX   
6019842  The Department of Housing Preservation and Dev...  MANHATTAN   

          Latitude  Longitude  
6019838  40.791359 -73.940184  
6019839  40.847809 -73.827481  
6019840  40.864764 -73.922764  
6019841  40.883207 -73.884279  
6019842  40.724749 -73.975269

In [5]:
nyc311.shape

(6019843, 16)

In [6]:
nyc311.columns

Index(['Unnamed: 0', 'Unique Key', 'Created Date', 'Closed Date',
       'Complaint Type', 'Location Type', 'Incident Zip', 'Incident Address',
       'Street Name', 'Address Type', 'City', 'Status',
       'Resolution Description', 'Borough', 'Latitude', 'Longitude'],
      dtype='object')

In [7]:
nyc311.dtypes

Unnamed: 0                  int64
Unique Key                  int64
Created Date               object
Closed Date                object
Complaint Type             object
Location Type              object
Incident Zip              float64
Incident Address           object
Street Name                object
Address Type               object
City                       object
Status                     object
Resolution Description     object
Borough                    object
Latitude                  float64
Longitude                 float64
dtype: object

In [8]:
Total_complaints=nyc311["Unique Key"].count()

In [9]:
Total_complaints

6019843

> Thus the nyc311 data set contains 6,019,843 rows i.e, *__6,019,843 complaints__*.There are 16 columns namely-*__'Unnamed: 0', 'Unique Key', 'Created Date', 'Closed Date','Complaint Type', 'Location Type', 'Incident Zip', 'Incident Address','Street Name', 'Address Type', 'City', 'Status','Resolution Description', 'Borough', 'Latitude' and 'Longitude'__*.
 The column 'Unnamed: 0' is unnecessary and hence can be dropped.The columns 'Created Date' and 'Closed Date' are of type object they need to be typecasted into type datetime.

In [10]:
nyc311.drop("Unnamed: 0", axis=1,inplace=True)

In [11]:
nyc311["Created Date"]=pd.to_datetime(nyc311["Created Date"],format="%m/%d/%Y %H:%M:%S %p")

In [12]:
nyc311["Closed Date"]=pd.to_datetime(nyc311["Closed Date"],format="%m/%d/%Y %H:%M:%S %p")

In [13]:
nyc311.dtypes

Unique Key                         int64
Created Date              datetime64[ns]
Closed Date               datetime64[ns]
Complaint Type                    object
Location Type                     object
Incident Zip                     float64
Incident Address                  object
Street Name                       object
Address Type                      object
City                              object
Status                            object
Resolution Description            object
Borough                           object
Latitude                         float64
Longitude                        float64
dtype: object

In [14]:
nyc311.head()

Unique Key        Created Date Closed Date        Complaint Type  \
0    45531130 2020-02-02 06:09:17         NaT        HEAT/HOT WATER   
1    45529784 2020-02-02 02:15:24         NaT  UNSANITARY CONDITION   
2    45527528 2020-02-02 02:27:41         NaT        HEAT/HOT WATER   
3    45530329 2020-02-02 12:13:18         NaT        HEAT/HOT WATER   
4    45528814 2020-02-02 01:59:44         NaT             APPLIANCE   

          Location Type  Incident Zip         Incident Address  \
0  RESIDENTIAL BUILDING       10019.0     426 WEST   52 STREET   
1  RESIDENTIAL BUILDING       11204.0           1751 67 STREET   
2  RESIDENTIAL BUILDING       11372.0          87-15 37 AVENUE   
3  RESIDENTIAL BUILDING       10458.0  2405 SOUTHERN BOULEVARD   
4  RESIDENTIAL BUILDING       11209.0            223 78 STREET   

          Street Name Address Type             City Status  \
0    WEST   52 STREET      ADDRESS         NEW YORK   Open   
1           67 STREET      ADDRESS         BROOKLYN   Open   
2           37 AVENUE      ADDRESS  Jackson Heights   Open   
3  SOUTHERN BOULEVARD      ADDRESS            BRONX   Open   
4           78 STREET      ADDRESS         BROOKLYN   Open   

                              Resolution Description    Borough   Latitude  \
0  The following complaint conditions are still o...  MANHATTAN  40.765132   
1  The following complaint conditions are still o...   BROOKLYN  40.618484   
2  The following complaint conditions are still o...     QUEENS  40.750269   
3  The following complaint conditions are still o...      BRONX  40.853773   
4  The following complaint conditions are still o...   BROOKLYN  40.629745   

   Longitude  
0 -73.988993  
1 -73.992673  
2 -73.879432  
3 -73.881558  
4 -74.030533

In [15]:
nyc311.tail()

Unique Key        Created Date         Closed Date  Complaint Type  \
6019838    44063685 2019-10-15 10:35:45 2019-10-16 05:55:02  HEAT/HOT WATER   
6019839    44063692 2019-10-15 06:09:28 2019-10-15 04:12:58  HEAT/HOT WATER   
6019840    44063724 2019-10-15 10:54:45 2019-10-16 06:14:29  HEAT/HOT WATER   
6019841    44063726 2019-10-15 02:07:34 2019-10-16 07:10:19  HEAT/HOT WATER   
6019842    44063737 2019-10-15 11:29:01 2019-10-15 09:19:45   PAINT/PLASTER   

                Location Type  Incident Zip         Incident Address  \
6019838  RESIDENTIAL BUILDING       10029.0     327 EAST  108 STREET   
6019839  RESIDENTIAL BUILDING       10461.0  3555 BRUCKNER BOULEVARD   
6019840  RESIDENTIAL BUILDING       10034.0       165 SHERMAN AVENUE   
6019841  RESIDENTIAL BUILDING       10467.0    66 WEST GUN HILL ROAD   
6019842  RESIDENTIAL BUILDING       10009.0             143 AVENUE D   

                Street Name Address Type      City  Status  \
6019838    EAST  108 STREET      ADDRESS  NEW YORK  Closed   
6019839  BRUCKNER BOULEVARD      ADDRESS     BRONX  Closed   
6019840      SHERMAN AVENUE      ADDRESS  NEW YORK  Closed   
6019841  WEST GUN HILL ROAD      ADDRESS     BRONX  Closed   
6019842            AVENUE D      ADDRESS  NEW YORK  Closed   

                                    Resolution Description    Borough  \
6019838  The complaint you filed is a duplicate of a co...  MANHATTAN   
6019839  The Department of Housing Preservation and Dev...      BRONX   
6019840  The complaint you filed is a duplicate of a co...  MANHATTAN   
6019841  The Department of Housing Preservation and Dev...      BRONX   
6019842  The Department of Housing Preservation and Dev...  MANHATTAN   

          Latitude  Longitude  
6019838  40.791359 -73.940184  
6019839  40.847809 -73.827481  
6019840  40.864764 -73.922764  
6019841  40.883207 -73.884279  
6019842  40.724749 -73.975269

> Arrange the data in descending order of "Created Date" i.e, the most recent complaint comes first in the data frame.

In [16]:
nyc311.sort_values(by=["Created Date"],inplace=True,ascending=False)

In [17]:
nyc311.head()

Unique Key        Created Date Closed Date  Complaint Type  \
394    45528930 2020-02-02 12:55:36         NaT  HEAT/HOT WATER   
173    45531005 2020-02-02 12:55:24         NaT  HEAT/HOT WATER   
645    45531855 2020-02-02 12:55:24         NaT        PLUMBING   
191    45531569 2020-02-02 12:53:07         NaT  HEAT/HOT WATER   
371    45528916 2020-02-02 12:52:27         NaT  HEAT/HOT WATER   

            Location Type  Incident Zip          Incident Address  \
394  RESIDENTIAL BUILDING       10025.0      216 WEST  104 STREET   
173  RESIDENTIAL BUILDING       10452.0      1000 ANDERSON AVENUE   
645  RESIDENTIAL BUILDING       11212.0         574 OSBORN STREET   
191  RESIDENTIAL BUILDING       11433.0  109-15 MERRICK BOULEVARD   
371  RESIDENTIAL BUILDING       11427.0    216-01 HILLSIDE AVENUE   

           Street Name Address Type            City Status  \
394   WEST  104 STREET      ADDRESS        NEW YORK   Open   
173    ANDERSON AVENUE      ADDRESS           BRONX   Open   
645      OSBORN STREET      ADDRESS        BROOKLYN   Open   
191  MERRICK BOULEVARD      ADDRESS         Jamaica   Open   
371    HILLSIDE AVENUE      ADDRESS  Queens Village   Open   

                                Resolution Description    Borough   Latitude  \
394  The complaint you filed is a duplicate of a co...  MANHATTAN  40.799406   
173  The following complaint conditions are still o...      BRONX  40.832418   
645  The following complaint conditions are still o...   BROOKLYN  40.657636   
191  The following complaint conditions are still o...     QUEENS  40.697151   
371  The following complaint conditions are still o...     QUEENS  40.727652   

     Longitude  
394 -73.967335  
173 -73.927059  
645 -73.905740  
191 -73.784159  
371 -73.748309

In [18]:
nyc311.tail()

Unique Key        Created Date         Closed Date  \
177180    15634558 2010-01-01 05:00:01 2010-01-04 01:20:48   
177188    15633410 2010-01-01 04:59:23 2010-01-02 05:02:49   
177186    15633862 2010-01-01 04:18:46 2010-01-04 01:20:46   
177179    15634698 2010-01-01 02:39:02 2010-01-04 01:20:45   
177182    15634460 2010-01-01 02:31:13 2010-01-02 02:35:01   

                Complaint Type Location Type  Incident Zip Incident Address  \
177180  HPD Literature Request           NaN           NaN              NaN   
177188  HPD Literature Request           NaN           NaN              NaN   
177186  HPD Literature Request           NaN           NaN              NaN   
177179  HPD Literature Request           NaN           NaN              NaN   
177182  HPD Literature Request           NaN           NaN              NaN   

       Street Name Address Type City  Status  \
177180         NaN          NaN  NaN  Closed   
177188         NaN          NaN  NaN  Closed   
177186         NaN          NaN  NaN  Closed   
177179         NaN          NaN  NaN  Closed   
177182         NaN          NaN  NaN  Closed   

                                   Resolution Description      Borough  \
177180  The literature has been mailed. If it has not ...  Unspecified   
177188  The literature will be emailed within 24 hours...  Unspecified   
177186  The literature has been mailed. If it has not ...  Unspecified   
177179  The literature has been mailed. If it has not ...  Unspecified   
177182  The literature will be emailed within 24 hours...  Unspecified   

        Latitude  Longitude  
177180       NaN        NaN  
177188       NaN        NaN  
177186       NaN        NaN  
177179       NaN        NaN  
177182       NaN        NaN

> Thus the nyc311 data set has complaints logged *__since 1st january,2010 till 2nd february,2019__*.Also we have mixed up the index, so we should reset the index

In [19]:
nyc311.reset_index(drop=True,inplace=True)

In [20]:
nyc311.head()

Unique Key        Created Date Closed Date  Complaint Type  \
0    45528930 2020-02-02 12:55:36         NaT  HEAT/HOT WATER   
1    45531005 2020-02-02 12:55:24         NaT  HEAT/HOT WATER   
2    45531855 2020-02-02 12:55:24         NaT        PLUMBING   
3    45531569 2020-02-02 12:53:07         NaT  HEAT/HOT WATER   
4    45528916 2020-02-02 12:52:27         NaT  HEAT/HOT WATER   

          Location Type  Incident Zip          Incident Address  \
0  RESIDENTIAL BUILDING       10025.0      216 WEST  104 STREET   
1  RESIDENTIAL BUILDING       10452.0      1000 ANDERSON AVENUE   
2  RESIDENTIAL BUILDING       11212.0         574 OSBORN STREET   
3  RESIDENTIAL BUILDING       11433.0  109-15 MERRICK BOULEVARD   
4  RESIDENTIAL BUILDING       11427.0    216-01 HILLSIDE AVENUE   

         Street Name Address Type            City Status  \
0   WEST  104 STREET      ADDRESS        NEW YORK   Open   
1    ANDERSON AVENUE      ADDRESS           BRONX   Open   
2      OSBORN STREET      ADDRESS        BROOKLYN   Open   
3  MERRICK BOULEVARD      ADDRESS         Jamaica   Open   
4    HILLSIDE AVENUE      ADDRESS  Queens Village   Open   

                              Resolution Description    Borough   Latitude  \
0  The complaint you filed is a duplicate of a co...  MANHATTAN  40.799406   
1  The following complaint conditions are still o...      BRONX  40.832418   
2  The following complaint conditions are still o...   BROOKLYN  40.657636   
3  The following complaint conditions are still o...     QUEENS  40.697151   
4  The following complaint conditions are still o...     QUEENS  40.727652   

   Longitude  
0 -73.967335  
1 -73.927059  
2 -73.905740  
3 -73.784159  
4 -73.748309

In [21]:
nyc311.tail()

Unique Key        Created Date         Closed Date  \
6019838    15634558 2010-01-01 05:00:01 2010-01-04 01:20:48   
6019839    15633410 2010-01-01 04:59:23 2010-01-02 05:02:49   
6019840    15633862 2010-01-01 04:18:46 2010-01-04 01:20:46   
6019841    15634698 2010-01-01 02:39:02 2010-01-04 01:20:45   
6019842    15634460 2010-01-01 02:31:13 2010-01-02 02:35:01   

                 Complaint Type Location Type  Incident Zip Incident Address  \
6019838  HPD Literature Request           NaN           NaN              NaN   
6019839  HPD Literature Request           NaN           NaN              NaN   
6019840  HPD Literature Request           NaN           NaN              NaN   
6019841  HPD Literature Request           NaN           NaN              NaN   
6019842  HPD Literature Request           NaN           NaN              NaN   

        Street Name Address Type City  Status  \
6019838         NaN          NaN  NaN  Closed   
6019839         NaN          NaN  NaN  Closed   
6019840         NaN          NaN  NaN  Closed   
6019841         NaN          NaN  NaN  Closed   
6019842         NaN          NaN  NaN  Closed   

                                    Resolution Description      Borough  \
6019838  The literature has been mailed. If it has not ...  Unspecified   
6019839  The literature will be emailed within 24 hours...  Unspecified   
6019840  The literature has been mailed. If it has not ...  Unspecified   
6019841  The literature has been mailed. If it has not ...  Unspecified   
6019842  The literature will be emailed within 24 hours...  Unspecified   

         Latitude  Longitude  
6019838       NaN        NaN  
6019839       NaN        NaN  
6019840       NaN        NaN  
6019841       NaN        NaN  
6019842       NaN        NaN

> Now we will study each of the attributes of the complaints.

In [22]:
nyc311["Complaint Type"].unique()

array(['HEAT/HOT WATER', 'PLUMBING', 'UNSANITARY CONDITION',
       'PAINT/PLASTER', 'DOOR/WINDOW', 'ELECTRIC', 'WATER LEAK',
       'GENERAL', 'APPLIANCE', 'SAFETY', 'OUTSIDE BUILDING', 'ELEVATOR',
       'FLOORING/STAIRS', 'Appliance', 'Mold', 'Unsanitary Condition',
       'Safety', 'Electric', 'General', 'Plumbing', 'Outside Building',
       'HPD Literature Request', 'AGENCY', 'VACANT APARTMENT', 'NONCONST',
       'HEATING', 'CONSTRUCTION', 'GENERAL CONSTRUCTION',
       'PAINT - PLASTER', 'STRUCTURAL'], dtype=object)

> We can see that there are duplicates in the values of column "Complaint Type".To solve this convert all of them into upper case.Also "HEAT/HOT WATER" and "HEATING" refers to the same type of complaint.Therefore replace it into "HEAT/HOT WATER".

In [23]:
nyc311["Complaint Type"]=nyc311["Complaint Type"].str.upper()

In [24]:
nyc311["Complaint Type"].replace("HEATING","HEAT/HOT WATER",inplace=True)

In [25]:
nyc311["Complaint Type"].replace("PAINT - PLASTER","PAINT/PLASTER",inplace=True)

In [26]:
comp_types=nyc311["Complaint Type"].value_counts().to_frame()

In [27]:
comp_types

Complaint Type
HEAT/HOT WATER                 2149424
PLUMBING                        711141
PAINT/PLASTER                   707695
GENERAL CONSTRUCTION            500863
UNSANITARY CONDITION            457142
ELECTRIC                        307311
NONCONST                        260890
DOOR/WINDOW                     205278
WATER LEAK                      193631
GENERAL                         152471
FLOORING/STAIRS                 137402
APPLIANCE                       112835
HPD LITERATURE REQUEST           52824
SAFETY                           51953
OUTSIDE BUILDING                  7148
ELEVATOR                          6725
CONSTRUCTION                      5078
STRUCTURAL                          16
AGENCY                               9
VACANT APARTMENT                     6
MOLD                                 1

In [28]:
comp_types.reset_index(inplace=True)

In [29]:
comp_types.rename(columns={"index":"Complaint Type","Complaint Type":"No of complaints"},inplace=True)

In [30]:
comp_types["Percentage"]=comp_types["No of complaints"]/Total_complaints*100

In [31]:
comp_types

Complaint Type  No of complaints  Percentage
0           HEAT/HOT WATER           2149424   35.705649
1                 PLUMBING            711141   11.813282
2            PAINT/PLASTER            707695   11.756037
3     GENERAL CONSTRUCTION            500863    8.320200
4     UNSANITARY CONDITION            457142    7.593919
5                 ELECTRIC            307311    5.104967
6                 NONCONST            260890    4.333834
7              DOOR/WINDOW            205278    3.410022
8               WATER LEAK            193631    3.216546
9                  GENERAL            152471    2.532807
10         FLOORING/STAIRS            137402    2.282485
11               APPLIANCE            112835    1.874384
12  HPD LITERATURE REQUEST             52824    0.877498
13                  SAFETY             51953    0.863029
14        OUTSIDE BUILDING              7148    0.118741
15                ELEVATOR              6725    0.111714
16            CONSTRUCTION              5078    0.084354
17              STRUCTURAL                16    0.000266
18                  AGENCY                 9    0.000150
19        VACANT APARTMENT                 6    0.000100
20                    MOLD                 1    0.000017

In [32]:
nyc311["Complaint Type"].isnull().value_counts()

False    6019843
Name: Complaint Type, dtype: int64

> Therefore, there are in total 21 complaint types."HEAT/HOT WATER" has the highest number of complaints, which is 2,149,424."MOLD" has the lowest number of complaints,which is 1.Also there is no complaint without a complaint type mentioned.

In [33]:
nyc311["Location Type"].unique()

array(['RESIDENTIAL BUILDING', 'Apartment', 'Building-Wide',
       'Public Area', 'Residential Building', nan], dtype=object)

In [34]:
nyc311["Location Type"]=nyc311["Location Type"].str.upper()

In [35]:
loc_types=nyc311["Location Type"].value_counts().to_frame()

In [36]:
loc_types

Location Type
RESIDENTIAL BUILDING        5967008
APARTMENT                         5
BUILDING-WIDE                     4
PUBLIC AREA                       2

In [37]:
loc_types.reset_index(inplace=True)

In [38]:
loc_types.rename(columns={"index":"location type","Location Type":"frequency"},inplace=True)

In [39]:
loc_types

location type  frequency
0  RESIDENTIAL BUILDING    5967008
1             APARTMENT          5
2         BUILDING-WIDE          4
3           PUBLIC AREA          2

In [40]:
nyc311["Location Type"].isnull().value_counts()

False    5967019
True       52824
Name: Location Type, dtype: int64

In [41]:
loc_types["Percentage"]=loc_types["frequency"]/Total_complaints*100

In [42]:
loc_types

location type  frequency  Percentage
0  RESIDENTIAL BUILDING    5967008   99.122319
1             APARTMENT          5    0.000083
2         BUILDING-WIDE          4    0.000066
3           PUBLIC AREA          2    0.000033

> Thus, there are more residential building complaints i.e, 5,967,008 and the least complaints are on public areas, which is just 2.Also there are 52,824 complaints whose location type is not mentioned.

In [43]:
nyc311["Incident Zip"].unique()

array([10025., 10452., 11212., 11433., 11427., 10040., 11238., 10453.,
       11216., 11378., 10457., 10029., 10456., 11385., 10016., 10474.,
       10458., 10003., 10466., 11203., 11372., 11104., 10462., 10031.,
       10024., 10463., 11435., 10075., 10026., 10009., 10035., 11210.,
       11208., 10039., 10027., 11432., 11236., 10468., 11434., 10459.,
       11373., 11369., 11226., 11230., 11214., 11219., 11207., 10470.,
       11368., 10472., 10034., 10473., 10011., 11201., 10314., 10467.,
       10014., 11213., 10471., 10033., 11237., 11224., 11692., 10460.,
       11204., 11221., 10019., 11231., 11102., 11235., 11218., 11412.,
       11229., 11101., 10455., 10030., 10304., 11356., 10028., 11228.,
       10461., 10022., 11225., 10036., 10302., 11377., 11355., 11354.,
       11419., 11217., 11423., 10454., 10032., 11417., 11413., 10301.,
       11691., 10010., 10002., 11374., 11233., 11211., 10128., 11232.,
       11223., 10023., 10037., 10451., 11375., 11106., 10307., 10469.,
      

In [44]:
len(nyc311["Incident Zip"].unique())

203

In [45]:
nyc311["Incident Zip"].isnull().value_counts()

False    5939146
True       80697
Name: Incident Zip, dtype: int64

> Thus, there are 203 different incident zip codes and 80,697 complaints has no incident zip code mentioned in it.

In [46]:
nyc311["Incident Address"].isnull().value_counts()

False    5967018
True       52825
Name: Incident Address, dtype: int64

> There are 52825 complaints with no incident address mentioned.

In [47]:
nyc311["Address Type"].unique()

array(['ADDRESS', nan], dtype=object)

In [48]:
nyc311.drop("Address Type",axis=1,inplace=True)

In [49]:
nyc311.columns

Index(['Unique Key', 'Created Date', 'Closed Date', 'Complaint Type',
       'Location Type', 'Incident Zip', 'Incident Address', 'Street Name',
       'City', 'Status', 'Resolution Description', 'Borough', 'Latitude',
       'Longitude'],
      dtype='object')

> The columns "Address Type" is not really necessary and hence is dropped.

In [50]:
nyc311["City"].unique()

array(['NEW YORK', 'BRONX', 'BROOKLYN', 'Jamaica', 'Queens Village',
       'Maspeth', 'Ridgewood', 'Jackson Heights', 'Sunnyside', 'Elmhurst',
       'East Elmhurst', 'Corona', 'STATEN ISLAND', 'Arverne', 'Astoria',
       'Saint Albans', 'Long Island City', 'College Point', 'Woodside',
       'Flushing', 'South Richmond Hill', 'Hollis', 'Ozone Park',
       'Springfield Gardens', 'Far Rockaway', 'Rego Park', 'Forest Hills',
       'Bayside', 'Richmond Hill', 'Rockaway Park', 'Woodhaven',
       'Floral Park', 'Fresh Meadows', 'Kew Gardens', 'Howard Beach',
       'Oakland Gardens', 'Middle Village', 'Cambria Heights', 'Rosedale',
       'South Ozone Park', 'Whitestone', nan, 'Little Neck', 'Bellerose',
       'New Hyde Park', 'Glen Oaks', 'QUEENS', 'Breezy Point',
       'JACKSON HEIGHTS', 'RICHMOND HILL', 'SAINT ALBANS', 'JAMAICA',
       'FAR ROCKAWAY', 'LONG ISLAND CITY', 'WOODSIDE', 'KEW GARDENS',
       'ASTORIA', 'CORONA', 'FRESH MEADOWS', 'RIDGEWOOD', 'FLUSHING',
       'QUEEN

In [51]:
nyc311["City"]=nyc311["City"].str.upper()

In [52]:
len(nyc311["City"].unique())

48

In [53]:
nyc311["City"].isnull().value_counts()

False    5939569
True       80274
Name: City, dtype: int64

In [54]:
city=nyc311["City"].value_counts().to_frame()

In [55]:
city.reset_index(inplace=True)

In [56]:
city.rename(columns={"index":"City","City":"Frequency"},inplace=True)

In [74]:
Total_complaints_with_city=city["Frequency"].sum(axis=0)

In [77]:
Total_complaints_with_city

5939569

In [80]:
city["Percentage"]=city["Frequency"]/Total_complaints_with_city*100

In [81]:
city

City  Frequency  Percentage
0              BROOKLYN    2026580   34.119984
1                 BRONX    1860522   31.324192
2              NEW YORK    1204417   20.277852
3         STATEN ISLAND     101553    1.709771
4               JAMAICA      97792    1.646449
5               ASTORIA      60604    1.020343
6              FLUSHING      59558    1.002733
7          FAR ROCKAWAY      51637    0.869373
8             RIDGEWOOD      47811    0.804957
9              ELMHURST      44037    0.741417
10               CORONA      34133    0.574671
11             WOODSIDE      30300    0.510138
12      JACKSON HEIGHTS      29661    0.499380
13         FOREST HILLS      20772    0.349722
14            REGO PARK      19230    0.323761
15            SUNNYSIDE      17274    0.290829
16       QUEENS VILLAGE      16146    0.271838
17  SOUTH RICHMOND HILL      15843    0.266737
18           OZONE PARK      15177    0.255524
19        RICHMOND HILL      14984    0.252274
20        EAST ELMHURST      13465    0.226700
21         SAINT ALBANS      13382    0.225303
22            WOODHAVEN      12963    0.218248
23          KEW GARDENS      12863    0.216565
24               HOLLIS      12591    0.211985
25              ARVERNE      12498    0.210419
26     SOUTH OZONE PARK      12161    0.204745
27  SPRINGFIELD GARDENS      11071    0.186394
28     LONG ISLAND CITY      10699    0.180131
29        ROCKAWAY PARK      10557    0.177740
30             ROSEDALE       7643    0.128679
31              MASPETH       6045    0.101775
32              BAYSIDE       4887    0.082279
33        FRESH MEADOWS       4733    0.079686
34         HOWARD BEACH       4351    0.073254
35       MIDDLE VILLAGE       4097    0.068978
36      OAKLAND GARDENS       4090    0.068860
37        COLLEGE POINT       3575    0.060190
38      CAMBRIA HEIGHTS       2885    0.048573
39           WHITESTONE       2206    0.037141
40          LITTLE NECK       1683    0.028335
41            BELLEROSE       1170    0.019698
42            GLEN OAKS        764    0.012863
43               QUEENS        491    0.008267
44          FLORAL PARK        287    0.004832
45         BREEZY POINT        269    0.004529
46        NEW HYDE PARK        112    0.001886

> There are more complaints on Brooklyn and the least on New Hyde Park.

In [59]:
nyc311["Status"].unique()

array(['Open', 'Closed', 'In Progress', 'Assigned', 'Pending'],
      dtype=object)

In [72]:
nyc311["Status"].isnull().value_counts()

False    6019843
Name: Status, dtype: int64

In [60]:
status=nyc311["Status"].value_counts().to_frame()

In [61]:
status.reset_index(inplace=True)

In [62]:
status.rename(columns={"index":"status","Status":"frequency"},inplace=True)

In [63]:
status["Percentage"]=status["frequency"]/Total_complaints*100

In [64]:
status

status  frequency  Percentage
0       Closed    5886253   97.780839
1         Open     133220    2.213015
2  In Progress        364    0.006047
3     Assigned          4    0.000066
4      Pending          2    0.000033

> 5,886,253 complaints are closed.

In [65]:
nyc311["Borough"].unique()

array(['MANHATTAN', 'BRONX', 'BROOKLYN', 'QUEENS', 'STATEN ISLAND',
       'Unspecified'], dtype=object)

In [70]:
nyc311["Borough"].isnull().value_counts()

False    6019843
Name: Borough, dtype: int64

In [66]:
borough=nyc311["Borough"].value_counts().to_frame()

In [67]:
borough.reset_index(inplace=True)

In [68]:
borough.rename(columns={"index":"borough","Borough":"frequency"},inplace=True)

In [69]:
borough["Percentage"]=borough["frequency"]/Total_complaints*100

In [71]:
borough

borough  frequency  Percentage
0       BROOKLYN    1739886   28.902515
1          BRONX    1617956   26.877046
2      MANHATTAN    1055225   17.529112
3    Unspecified     873221   14.505711
4         QUEENS     645971   10.730695
5  STATEN ISLAND      87584    1.454922

> Brooklyn has the highest number of complaints, followed by Bronx. Staten Island has the least number of complaints.

In [73]:
heat_hotwater=nyc311.groupby("Complaint Type").get_group("HEAT/HOT WATER")

In [82]:
heat_hotwater.head()

Unique Key        Created Date Closed Date  Complaint Type  \
0    45528930 2020-02-02 12:55:36         NaT  HEAT/HOT WATER   
1    45531005 2020-02-02 12:55:24         NaT  HEAT/HOT WATER   
3    45531569 2020-02-02 12:53:07         NaT  HEAT/HOT WATER   
4    45528916 2020-02-02 12:52:27         NaT  HEAT/HOT WATER   
7    45528479 2020-02-02 12:44:07         NaT  HEAT/HOT WATER   

          Location Type  Incident Zip          Incident Address  \
0  RESIDENTIAL BUILDING       10025.0      216 WEST  104 STREET   
1  RESIDENTIAL BUILDING       10452.0      1000 ANDERSON AVENUE   
3  RESIDENTIAL BUILDING       11433.0  109-15 MERRICK BOULEVARD   
4  RESIDENTIAL BUILDING       11427.0    216-01 HILLSIDE AVENUE   
7  RESIDENTIAL BUILDING       10453.0      111 WEST  183 STREET   

         Street Name            City Status  \
0   WEST  104 STREET        NEW YORK   Open   
1    ANDERSON AVENUE           BRONX   Open   
3  MERRICK BOULEVARD         JAMAICA   Open   
4    HILLSIDE AVENUE  QUEENS VILLAGE   Open   
7   WEST  183 STREET           BRONX   Open   

                              Resolution Description    Borough   Latitude  \
0  The complaint you filed is a duplicate of a co...  MANHATTAN  40.799406   
1  The following complaint conditions are still o...      BRONX  40.832418   
3  The following complaint conditions are still o...     QUEENS  40.697151   
4  The following complaint conditions are still o...     QUEENS  40.727652   
7  The following complaint conditions are still o...      BRONX  40.860289   

   Longitude  
0 -73.967335  
1 -73.927059  
3 -73.784159  
4 -73.748309  
7 -73.909115

In [83]:
hhw_borough=heat_hotwater["Borough"].value_counts().to_frame()

In [84]:
hhw_borough

Borough
BRONX           606133
BROOKLYN        574845
MANHATTAN       423039
Unspecified     282917
QUEENS          244122
STATEN ISLAND    18368

In [85]:
hhw_borough.reset_index(inplace=True)

In [86]:
hhw_borough.rename(columns={"index":"Borough","Borough":"frequency"},inplace=True)

In [87]:
total_hhw_complaints=hhw_borough["frequency"].sum(axis=0)

In [88]:
total_hhw_complaints

2149424

In [89]:
hhw_borough["percentage"]=hhw_borough["frequency"]/total_hhw_complaints*100

In [90]:
hhw_borough

Borough  frequency  percentage
0          BRONX     606133   28.199787
1       BROOKLYN     574845   26.744142
2      MANHATTAN     423039   19.681505
3    Unspecified     282917   13.162457
4         QUEENS     244122   11.357554
5  STATEN ISLAND      18368    0.854555

In [91]:
heat_hotwater["Incident Zip"].value_counts()

11226.0    69041
10467.0    66073
10458.0    65372
10468.0    58190
10453.0    57818
           ...  
10103.0        1
10112.0        1
10121.0        1
10129.0        1
12345.0        1
Name: Incident Zip, Length: 193, dtype: int64

In [92]:
hhw_address=heat_hotwater["Incident Address"].value_counts().to_frame()

In [93]:
hhw_address.reset_index(inplace=True)

In [94]:
hhw_address.rename(columns={"index":"address","Incident Address":"frequency"},inplace=True)

In [95]:
hhw_address

address  frequency
0             34 ARDEN STREET      13858
1       89-21 ELMHURST AVENUE      11548
2          3810 BAILEY AVENUE       7115
3         1025 BOYNTON AVENUE       5785
4             9511 SHORE ROAD       5006
...                       ...        ...
136944           366 5 STREET          1
136945       817 UNION STREET          1
136946     510 KAPPOCK STREET          1
136947   325 EAST   10 STREET          1
136948         1742 80 STREET          1

[136949 rows x 2 columns]

In [96]:
hhw_address["percentage"]=hhw_address["frequency"]/total_hhw_complaints

In [97]:
hhw_address

address  frequency    percentage
0             34 ARDEN STREET      13858  6.447309e-03
1       89-21 ELMHURST AVENUE      11548  5.372602e-03
2          3810 BAILEY AVENUE       7115  3.310189e-03
3         1025 BOYNTON AVENUE       5785  2.691419e-03
4             9511 SHORE ROAD       5006  2.328996e-03
...                       ...        ...           ...
136944           366 5 STREET          1  4.652409e-07
136945       817 UNION STREET          1  4.652409e-07
136946     510 KAPPOCK STREET          1  4.652409e-07
136947   325 EAST   10 STREET          1  4.652409e-07
136948         1742 80 STREET          1  4.652409e-07

[136949 rows x 3 columns]

In [98]:
hhw_bronx=heat_hotwater.groupby("Borough").get_group("BRONX")

In [114]:
hhw_bronx.head()

Unique Key        Created Date         Closed Date  Complaint Type  \
1     45531005 2020-02-02 12:55:24                 NaT  HEAT/HOT WATER   
7     45528479 2020-02-02 12:44:07                 NaT  HEAT/HOT WATER   
11    45527545 2020-02-02 12:38:05 2020-02-02 06:53:37  HEAT/HOT WATER   
17    45530314 2020-02-02 12:33:05                 NaT  HEAT/HOT WATER   
18    45528473 2020-02-02 12:30:51                 NaT  HEAT/HOT WATER   

           Location Type  Incident Zip      Incident Address  \
1   RESIDENTIAL BUILDING       10452.0  1000 ANDERSON AVENUE   
7   RESIDENTIAL BUILDING       10453.0  111 WEST  183 STREET   
11  RESIDENTIAL BUILDING       10457.0   2176 TIEBOUT AVENUE   
17  RESIDENTIAL BUILDING       10474.0    1314 SENECA AVENUE   
18  RESIDENTIAL BUILDING       10452.0   1307 MERRIAM AVENUE   

         Street Name   City  Status  \
1    ANDERSON AVENUE  BRONX    Open   
7   WEST  183 STREET  BRONX    Open   
11    TIEBOUT AVENUE  BRONX  Closed   
17     SENECA AVENUE  BRONX    Open   
18    MERRIAM AVENUE  BRONX    Open   

                               Resolution Description Borough   Latitude  \
1   The following complaint conditions are still o...   BRONX  40.832418   
7   The following complaint conditions are still o...   BRONX  40.860289   
11  The Department of Housing Preservation and Dev...   BRONX  40.854439   
17  The following complaint conditions are still o...   BRONX  40.818939   
18  The following complaint conditions are still o...   BRONX  40.840709   

    Longitude  
1  -73.927059  
7  -73.909115  
11 -73.898423  
17 -73.886981  
18 -73.926150

In [99]:
hhw_bronx["Incident Zip"].unique()

array([10452., 10453., 10457., 10474., 10466., 10462., 10458., 10468.,
       10472., 10467., 10471., 10460., 10463., 10455., 10456., 10473.,
       10461., 10454., 10470., 10459., 10451., 10469., 10465., 10475.,
       10464.,    nan, 10803.])

In [100]:
hhw_bronx["Incident Zip"].isnull().value_counts()

False    600147
True       5986
Name: Incident Zip, dtype: int64

In [101]:
hhw_bronx["Incident Zip"].value_counts()

10467.0    57190
10458.0    56992
10468.0    50670
10453.0    49938
10452.0    43309
10457.0    40648
10456.0    38073
10462.0    35322
10463.0    30185
10472.0    29796
10460.0    28130
10451.0    23186
10466.0    21170
10459.0    19937
10455.0    15139
10461.0    14098
10454.0     9719
10469.0     9054
10473.0     8625
10470.0     6911
10474.0     5144
10471.0     3616
10465.0     2157
10475.0      794
10464.0      340
10803.0        4
Name: Incident Zip, dtype: int64

In [117]:
req_hhw_bronx=hhw_bronx["Incident Address"].value_counts().to_frame()

In [118]:
req_hhw_bronx

Incident Address
3810 BAILEY AVENUE                    7115
1025 BOYNTON AVENUE                   5774
2090 EAST TREMONT AVENUE              3710
888 GRAND CONCOURSE                   3034
2968 PERRY AVENUE                     2925
...                                    ...
1410 BRONX RIVER AVENUE                  1
738 SOUTHERN BOULEVARD                   1
4429 EDSON AVENUE                        1
630 EAST 187 STREET                      1
2445 OLINVILLE AVENUE                    1

[22859 rows x 1 columns]

In [119]:
req_hhw_bronx.reset_index(inplace=True)

In [121]:
req_hhw_bronx

index  Incident Address
0            3810 BAILEY AVENUE              7115
1           1025 BOYNTON AVENUE              5774
2      2090 EAST TREMONT AVENUE              3710
3           888 GRAND CONCOURSE              3034
4             2968 PERRY AVENUE              2925
...                         ...               ...
22854   1410 BRONX RIVER AVENUE                 1
22855    738 SOUTHERN BOULEVARD                 1
22856         4429 EDSON AVENUE                 1
22857       630 EAST 187 STREET                 1
22858     2445 OLINVILLE AVENUE                 1

[22859 rows x 2 columns]

In [125]:
req_hhw_bronx.rename(columns={"index":"Address","Incident Address":"Frequency"},inplace=True)

In [126]:
req_hhw_bronx.head()

Address  No of complaints
0        3810 BAILEY AVENUE              7115
1       1025 BOYNTON AVENUE              5774
2  2090 EAST TREMONT AVENUE              3710
3       888 GRAND CONCOURSE              3034
4         2968 PERRY AVENUE              2925

In [103]:
heat_hotwater["Status"].value_counts()

Closed     2133331
Open         16091
Pending          2
Name: Status, dtype: int64

In [104]:
col_list=["Block","Address","BldgArea","BldgDepth","BuiltFAR","CommFAR","FacilFAR","Lot","LotArea","LotDepth","NumBldgs","NumFloors","OfficeArea","ResArea","ResidFAR","RetailArea","YearBuilt","YearAlter1","ZipCode","YCoord","XCoord"]
bronx_data=pd.read_csv(r"C:\Users\Josna Baby\Downloads\BX_18v1.csv",usecols=col_list)

In [105]:
bronx_data.head()

Block  Lot  ZipCode                 Address  LotArea  BldgArea  ResArea  \
0   2260    1  10454.0  122 BRUCKNER BOULEVARD    15000         0        0   
1   2260    4  10454.0  126 BRUCKNER BOULEVARD    13770       752        0   
2   2260   10  10454.0  138 BRUCKNER BOULEVARD    35000     39375        0   
3   2260   17  10454.0  144 BRUCKNER BOULEVARD     2500     12500    12500   
4   2260   18  10454.0  148 BRUCKNER BOULEVARD     1875      8595     6876   

   OfficeArea  RetailArea  NumBldgs  ...  LotDepth  BldgDepth  YearBuilt  \
0           0           0         1  ...     200.0        0.0          0   
1         272           0         2  ...     100.0       16.0       1931   
2           0           0         1  ...     200.0      200.0       1931   
3           0           0         1  ...     100.0       85.0       1931   
4           0        1719         1  ...      75.0       70.0       1920   

   YearAlter1  BuiltFAR  ResidFAR  CommFAR  FacilFAR     XCoord    YCoord  
0           0      0.00      6.02      5.0       6.5  1005957.0  232162.0  
1        1994      0.05      6.02      5.0       6.5  1006076.0  232156.0  
2           0      1.13      6.02      5.0       6.5  1006187.0  232036.0  
3        2001      5.00      6.02      5.0       6.5  1006299.0  232033.0  
4        2009      4.58      6.02      5.0       6.5  1006363.0  232040.0  

[5 rows x 21 columns]

In [106]:
bronx_data.tail()

Block   Lot  ZipCode Address  LotArea  BldgArea  ResArea  OfficeArea  \
89849   9999   100      NaN     NaN        0         0        0           0   
89850   9999   150      NaN     NaN        0         0        0           0   
89851   9999   200      NaN     NaN        0         0        0           0   
89852   9999  8900      NaN     NaN        0         0        0           0   
89853      0  8900      NaN     NaN        0         0        0           0   

       RetailArea  NumBldgs  ...  LotDepth  BldgDepth  YearBuilt  YearAlter1  \
89849           0         0  ...       0.0        0.0          0           0   
89850           0         0  ...       0.0        0.0          0           0   
89851           0         0  ...       0.0        0.0          0           0   
89852           0         0  ...       0.0        0.0          0           0   
89853           0         0  ...       0.0        0.0          0           0   

       BuiltFAR  ResidFAR  CommFAR  FacilFAR  XCoord  YCoord  
89849       0.0      0.00      0.0       0.0     NaN     NaN  
89850       0.0      6.02      3.4       6.5     NaN     NaN  
89851       0.0      0.00      0.0       0.0     NaN     NaN  
89852       0.0      0.00      0.0       0.0     NaN     NaN  
89853       0.0      0.00      0.0       0.0     NaN     NaN  

[5 rows x 21 columns]

In [96]:
bronx_data.columns

Index(['Block', 'Lot', 'ZipCode', 'Address', 'LotArea', 'BldgArea', 'ResArea',
       'OfficeArea', 'RetailArea', 'NumBldgs', 'NumFloors', 'LotDepth',
       'BldgDepth', 'YearBuilt', 'YearAlter1', 'BuiltFAR', 'ResidFAR',
       'CommFAR', 'FacilFAR', 'XCoord', 'YCoord'],
      dtype='object')

In [127]:
array=req_hhw_bronx["Address"]

In [128]:
array

0              3810 BAILEY AVENUE
1             1025 BOYNTON AVENUE
2        2090 EAST TREMONT AVENUE
3             888 GRAND CONCOURSE
4               2968 PERRY AVENUE
                   ...           
22854     1410 BRONX RIVER AVENUE
22855      738 SOUTHERN BOULEVARD
22856           4429 EDSON AVENUE
22857         630 EAST 187 STREET
22858       2445 OLINVILLE AVENUE
Name: Address, Length: 22859, dtype: object

In [129]:
req_data=bronx_data.loc[bronx_data["Address"].isin(array)]

In [130]:
req_data.head()

Block  Lot  ZipCode                 Address  LotArea  BldgArea  ResArea  \
3    2260   17  10454.0  144 BRUCKNER BOULEVARD     2500     12500    12500   
4    2260   18  10454.0  148 BRUCKNER BOULEVARD     1875      8595     6876   
28   2261   45  10454.0  147 BRUCKNER BOULEVARD     2500     11500     9176   
45   2263   72  10454.0        169 BROOK AVENUE     1700      2080     2080   
46   2263   73  10454.0        167 BROOK AVENUE     1700      2080     2080   

    OfficeArea  RetailArea  NumBldgs  ...  LotDepth  BldgDepth  YearBuilt  \
3            0           0         1  ...    100.00       85.0       1931   
4            0        1719         1  ...     75.00       70.0       1920   
28           0        2324         1  ...    100.00       96.0       1925   
45           0           0         1  ...     85.06       52.0       2000   
46           0           0         1  ...     85.06       52.0       2000   

    YearAlter1  BuiltFAR  ResidFAR  CommFAR  FacilFAR     XCoord    YCoord  
3         2001      5.00      6.02      5.0       6.5  1006299.0  232033.0  
4         2009      4.58      6.02      5.0       6.5  1006363.0  232040.0  
28           0      4.60      3.00      2.0       3.0  1006412.0  232154.0  
45           0      1.22      2.43      0.0       4.8  1006189.0  232909.0  
46           0      1.22      2.43      0.0       4.8  1006179.0  232892.0  

[5 rows x 21 columns]

In [131]:
req_data.reset_index(inplace=True,drop=True)

In [ ]:
req_data["no of complaints"]=[]
for i in req_data["Address"]:
    req_data["no of complaints"]=req_data["no of complaints"].append()
    

In [132]:
req_data.head()

Block  Lot  ZipCode                 Address  LotArea  BldgArea  ResArea  \
0   2260   17  10454.0  144 BRUCKNER BOULEVARD     2500     12500    12500   
1   2260   18  10454.0  148 BRUCKNER BOULEVARD     1875      8595     6876   
2   2261   45  10454.0  147 BRUCKNER BOULEVARD     2500     11500     9176   
3   2263   72  10454.0        169 BROOK AVENUE     1700      2080     2080   
4   2263   73  10454.0        167 BROOK AVENUE     1700      2080     2080   

   OfficeArea  RetailArea  NumBldgs  ...  LotDepth  BldgDepth  YearBuilt  \
0           0           0         1  ...    100.00       85.0       1931   
1           0        1719         1  ...     75.00       70.0       1920   
2           0        2324         1  ...    100.00       96.0       1925   
3           0           0         1  ...     85.06       52.0       2000   
4           0           0         1  ...     85.06       52.0       2000   

   YearAlter1  BuiltFAR  ResidFAR  CommFAR  FacilFAR     XCoord    YCoord  
0        2001      5.00      6.02      5.0       6.5  1006299.0  232033.0  
1        2009      4.58      6.02      5.0       6.5  1006363.0  232040.0  
2           0      4.60      3.00      2.0       3.0  1006412.0  232154.0  
3           0      1.22      2.43      0.0       4.8  1006189.0  232909.0  
4           0      1.22      2.43      0.0       4.8  1006179.0  232892.0  

[5 rows x 21 columns]

In [97]:
bronx_data.dtypes

Block           int64
Lot             int64
ZipCode       float64
Address        object
LotArea         int64
BldgArea        int64
ResArea         int64
OfficeArea      int64
RetailArea      int64
NumBldgs        int64
NumFloors     float64
LotDepth      float64
BldgDepth     float64
YearBuilt       int64
YearAlter1      int64
BuiltFAR      float64
ResidFAR      float64
CommFAR       float64
FacilFAR      float64
XCoord        float64
YCoord        float64
dtype: object

In [98]:
bronx_data["YearBuilt"].isnull().value_counts()

False    89854
Name: YearBuilt, dtype: int64

In [99]:
bronx_data.shape

(89854, 21)

In [100]:
bronx_data["Lot"].value_counts()

1       2486
20      1207
15      1178
5       1169
25      1168
        ... 
1636       1
3813       1
1892       1
2148       1
2049       1
Name: Lot, Length: 1960, dtype: int64

In [101]:
import numpy as np

In [102]:
bronx_data[np.logical_and(bronx_data["Lot"]==1,bronx_data["Block"]==2260)]

Block  Lot  ZipCode                 Address  LotArea  BldgArea  ResArea  \
0   2260    1  10454.0  122 BRUCKNER BOULEVARD    15000         0        0   

   OfficeArea  RetailArea  NumBldgs  ...  LotDepth  BldgDepth  YearBuilt  \
0           0           0         1  ...     200.0        0.0          0   

   YearAlter1  BuiltFAR  ResidFAR  CommFAR  FacilFAR     XCoord    YCoord  
0           0       0.0      6.02      5.0       6.5  1005957.0  232162.0  

[1 rows x 21 columns]

In [103]:
hhw_pluto_bronx=pd.concat([bronx_data.groupby("ZipCode").get_group(i) for i in [10452.,10453., 10457.,10474., 10466., 10462., 10458., 10468.,
       10472., 10467., 10471., 10460., 10463., 10455., 10456., 10473.,
       10461., 10454., 10470., 10459., 10451., 10469., 10465., 10475.,
       10464.]])

In [104]:
hhw_pluto_bronx.shape

(89523, 21)

In [105]:
hhw_pluto_bronx.reset_index(drop=True,inplace=True)

In [106]:
hhw_pluto_bronx.head()

Block  Lot  ZipCode               Address  LotArea  BldgArea  ResArea  \
0   2463    1  10452.0    1170 WALTON AVENUE    12500     57500    57500   
1   2463    6  10452.0    1176 WALTON AVENUE    12500     57500    57500   
2   2463   11  10452.0       112 TUDOR PLACE    13686     57240    57240   
3   2463   16  10452.0  1173 GRAND CONCOURSE    13569     48395    46747   
4   2463   21  10452.0  1153 GRAND CONCOURSE    13713     59300    58300   

   OfficeArea  RetailArea  NumBldgs  ...  LotDepth  BldgDepth  YearBuilt  \
0           0           0         1  ...     100.0      100.0       1922   
1           0           0         1  ...     100.0      100.0       1922   
2           0           0         1  ...     125.0      106.0       1922   
3         500        1148         1  ...     120.0      110.0       1922   
4           0        1000         1  ...     100.0       90.0       1922   

   YearAlter1  BuiltFAR  ResidFAR  CommFAR  FacilFAR     XCoord    YCoord  
0        1991      4.60      6.02      0.0       6.5  1006397.0  243022.0  
1        1991      4.60      6.02      0.0       6.5  1006448.0  243138.0  
2        1991      4.18      6.02      0.0       6.5  1006545.0  243094.0  
3           0      3.57      6.02      0.0       6.5  1006649.0  243050.0  
4           0      4.32      6.02      0.0       6.5  1006577.0  242943.0  

[5 rows x 21 columns]

In [107]:
hhw_pluto_bronx.tail()

Block  Lot  ZipCode            Address  LotArea  BldgArea  ResArea  \
89518   5652  198  10464.0         PARK DRIVE    10000         0        0   
89519   5652  203  10464.0         PARK DRIVE     1800         0        0   
89520   5652  215  10464.0         PARK DRIVE     3188         0        0   
89521   5652  222  10464.0      11 PARK DRIVE    17250      3350     3350   
89522   5652  224  10464.0  2 SPLIT ROCK ROAD     9172      6574     6574   

       OfficeArea  RetailArea  NumBldgs  ...  LotDepth  BldgDepth  YearBuilt  \
89518           0           0         0  ...    100.00        0.0          0   
89519           0           0         0  ...    100.00        0.0          0   
89520           0           0         0  ...    115.58        0.0          0   
89521           0           0         1  ...    113.00       44.5       1975   
89522           0           0         2  ...    113.00       36.0       1970   

       YearAlter1  BuiltFAR  ResidFAR  CommFAR  FacilFAR     XCoord    YCoord  
89518           0      0.00       0.5      0.0       1.0        NaN       NaN  
89519           0      0.00       0.5      0.0       1.0        NaN       NaN  
89520           0      0.00       0.5      0.0       1.0        NaN       NaN  
89521           0      0.19       0.5      0.0       1.0  1036012.0  262947.0  
89522        2003      0.72       0.5      0.0       1.0  1035870.0  262997.0  

[5 rows x 21 columns]

In [108]:
hhw_pluto_bronx["NumFloors"].value_counts().to_frame()

NumFloors
2.00       42186
3.00       15206
1.00        9021
0.00        6234
2.50        5077
5.00        3239
6.00        2171
4.00        1710
1.50        1531
2.75        1165
1.75         716
1.67         341
7.00         233
8.00         160
9.00          77
3.50          61
12.00         51
10.00         43
14.00         42
11.00         39
13.00         38
20.00         25
16.00         24
2.67          20
3.75          18
19.00         13
21.00         12
15.00         11
18.00         11
2.25           9
22.00          7
25.00          6
30.00          5
17.00          5
28.00          2
33.00          2
23.00          2
31.00          1
29.00          1
5.50           1
1.33           1
0.50           1
4.50           1
24.00          1
44.00          1
1.25           1
42.00          1

In [109]:
hhw_pluto_bronx["NumFloors"]=round(hhw_pluto_bronx["NumFloors"])

In [110]:
hhw_pluto_bronx["NumFloors"].value_counts()

2.0     49860
3.0     16391
1.0      9023
0.0      6235
5.0      3239
6.0      2172
4.0      1790
7.0       233
8.0       160
9.0        77
12.0       51
10.0       43
14.0       42
11.0       39
13.0       38
20.0       25
16.0       24
19.0       13
21.0       12
18.0       11
15.0       11
22.0        7
25.0        6
30.0        5
17.0        5
33.0        2
28.0        2
23.0        2
24.0        1
29.0        1
42.0        1
44.0        1
31.0        1
Name: NumFloors, dtype: int64

In [ ]:
from scipy import stats
f_value,p_value=stats.f_oneway("NumFloors")